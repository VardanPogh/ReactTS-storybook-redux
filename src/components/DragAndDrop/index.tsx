import React from 'react'
import styled from '@emotion/styled'
import { makeStyles } from '@material-ui/core/styles'
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd'
import { Grid } from '@material-ui/core'

const reorder = (list: any, startIndex: number, endIndex: number): any => {
  const result = Array.from(list)
  const [removed] = result.splice(startIndex, 1)
  result.splice(endIndex, 0, removed)
  return result
}

const BoxItem = styled.div`
  background: #ffffff 0% 0% no-repeat padding-box;
  box-shadow: 0px 0px 6px #00000014;
  border: 1px dashed #c4c4c4;
  height: auto;
  margin-left: 32px;
  padding: 20px 36px;
  box-sizing: border-box;
  margin-bottom: 30px;
  font-size: 18px;
  font-family: 'Poppins, sans-serif';
  color: #014bad;
  text-align: center;
  width: 65%;
  @media (max-width: 375px) {
    width: 100%;
  }
`

const useStyles = makeStyles(() => ({
  numbers: {
    fontSize: 77,
    color: '#545454',
    opacity: 0.3,
    listStyle: 'none',
    marginTop: 0,
    paddingLeft: 0,
    '& li': {
      listStyle: 'none',
      lineHeight: 1,
      marginBottom: 18,
    },
  },
}))
interface Box {
  box: any
  index: number
}

const BoxComponent: React.FC<Box> = ({ box, index }) => {
  return (
    <Draggable draggableId={box.id} index={index}>
      {(provided): React.ReactElement => (
        <BoxItem
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
        >
          {box.content}
        </BoxItem>
      )}
    </Draggable>
  )
}
const BoxList = React.memo(function BoxList({ boxes }: any) {
  return boxes.map((box: any, index: number) => (
    <BoxComponent box={box} index={index} key={box.id} />
  ))
})

interface Props {
  values: any
  setValues: any
}
const DragAndDrop: React.FC<Props> = ({ values, setValues }) => {
  const classes = useStyles()
  const onDragEnd = (result: any): void => {
    if (!result.destination) {
      return
    }
    if (result.destination.index === result.source.index) {
      return
    }
    const boxes = reorder(
      values.boxes,
      result.source.index,
      result.destination.index
    )
    setValues({ boxes })
  }
  const LI = values.boxes.map((item: any, index: number) => (
    <li key={index}>{index + 1}</li>
  ))

  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <Grid container>
        <Grid item xs={false} sm={2} />
        <Grid item xs={1} sm={1}>
          <ul className={classes.numbers}>{LI}</ul>
        </Grid>
        <Grid item xs={8} sm={8}>
          <Droppable droppableId="list">
            {(provided): React.ReactElement => (
              <div ref={provided.innerRef} {...provided.droppableProps}>
                <BoxList boxes={values.boxes} />
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        </Grid>
        <Grid item xs={false} sm={2} />
      </Grid>
    </DragDropContext>
  )
}

export default DragAndDrop
