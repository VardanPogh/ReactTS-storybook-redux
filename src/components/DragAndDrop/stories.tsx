import React, { useState } from 'react'
import DragAndDrop from '.'

const List = [
  {
    id: 'box-0',
    content: `Maximise Performance`,
  },
  {
    id: 'box-1',
    content: `Minimise Life Cycle Cost`,
  },
  {
    id: 'box-2',
    content: `Minimise Quality Risk`,
  },
]

export default {
  title: 'DragAndDrop',
  component: DragAndDrop,
}
export const Boxes: React.FC = () => {
  const [state, setState] = useState({ boxes: List })
  // console log to check corrrect values will be passed to form
  // console.log(state)
  return <DragAndDrop values={state} setValues={setState} />
}
