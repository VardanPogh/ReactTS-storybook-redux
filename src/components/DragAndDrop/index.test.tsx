import React from 'react'
import { render } from '@testing-library/react'
import DragAndDrop from '.'

const List = [
  {
    id: 'box-0',
    content: `Maximise Performance`,
  },
  {
    id: 'box-1',
    content: `Minimise Life Cycle Cost`,
  },
  {
    id: 'box-2',
    content: `Minimise Quality Risk`,
  },
]

describe('DragAndDrop Test', () => {
  test('Load DragAndDrop', () => {
    const myMock = jest.fn()
    const { container } = render(
      <DragAndDrop values={{ boxes: List }} setValues={myMock} />
    )
    expect(container).toMatchSnapshot()
  })
})
