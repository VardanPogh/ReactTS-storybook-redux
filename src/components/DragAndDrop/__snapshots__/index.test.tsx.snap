// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragAndDrop Test Load DragAndDrop 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-2"
    />
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-1"
    >
      <ul
        class="makeStyles-numbers-1"
      >
        <li>
          1
        </li>
        <li>
          2
        </li>
        <li>
          3
        </li>
      </ul>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 MuiGrid-grid-sm-8"
    >
      <div
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="list"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          class="css-1jk3y74"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="box-0"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="box-0"
          draggable="false"
          role="button"
          tabindex="0"
        >
          Maximise Performance
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          class="css-1jk3y74"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="box-1"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="box-1"
          draggable="false"
          role="button"
          tabindex="0"
        >
          Minimise Life Cycle Cost
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          class="css-1jk3y74"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="box-2"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="box-2"
          draggable="false"
          role="button"
          tabindex="0"
        >
          Minimise Quality Risk
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-2"
    />
  </div>
</div>
`;
